mantidproject/mantid
# New versions of `ReflectometryReductionOne` and `CreateTransmissionWorkspace`

`ReflectometryReductionOne` and `CreateTransmissionWorkspace` (see [here](http://docs.mantidproject.org/nightly/algorithms/ReflectometryReductionOne-v1.html) and [here](http://docs.mantidproject.org/nightly/algorithms/CreateTransmissionWorkspace-v1.html) for the documentation) have some duplicate steps: for instance, in `ReflectometryReductionOne`, `ConvertUnits` is applied three times to the input workspace, and it is applied twice in the second algorithm. In addition, `ReflectometryReductionOne` has some known problems, the main ones:
- Normalization by monitors is optional. However, when no monitors are specified, given the current structure of the algorithm, we have to create a workspace filled with `1`s to normalize the detector ws. This should not be necessary.
- Even if no monitors are specified, if the property `NormalizeByIntegratedMonitors` (which is `True` by default) is not manually set to `False` the algorithm will use the workspace filled with `1` as a monitor ws, integrate it, and use it to divide detectors by monitors. This shouldn't happen. If no monitors are specified we shouldn't need to set this property to `false`, it should just not be taken into account.
- The algorithm has a property `CorrectDetectorPositions` which is set to `True` by default and essentially moves the detectors to the correct position according to the angle specified by the user. In adition, the algorithm, moves the source according to the same angle. However, when the user sets `CorrectDetectorPositions` to `False`, the algorithm will still move the source, which will cause `ConvertUnits` to produce wrong `Q` values in the output workspace.

`CreateTransmissionWorkspace` also has some issues:
- According to the documentation, the only mandatory properties are `ProcessingInstructions`, `WavelengthMin` and `WavelengthMax`. However, trying to run the algorithm providing only those properties will result in an error `A value must be entered for this parameter`. The problem seems to be that the algorithm is trying to run `CalculateFlatBackground` to subtract the background from the monitors, even if we didn't specify any monitors.

This PR adds new versions for both algorithms, fixing the above issues and hopefully arraging the code in a way that will make it easier to add new input properties in the future.

**To test:**

- Make sure that algorithms `ReflectometryReductionOneAuto` and `CreateTransmissionWorkspaceAuto` are using version 1 of `ReflectometryReductionOne` and `CreateTransmissionWorkspace`.
- Make sure that `ReflectometryReductionOneTest` is testing version 1 of `ReflectometryReductionOne` and `CreateTransmissionWorkspaceTest` is testing version 1 of `CreateTransmissionWorkspace`.
- Build the documentation and make sure that the user examples in `ReflectometryReductionOne-v1` correspond to version 1 (the same for `CreateTransmissionWorkspace-v1`).
- Build the documentation and make sure that `ReflectometryReductionOne-v2` and `CreateTransmissionWorkspace-v2` have a note about previous versions.
- Compare resulst from versions 1 and 2. I have prepared two scripts to do this, see my comments below. Run the scripts and make sure that both versions produce results that are compatible within error bars.

Fixes #17429.

Release notes were updated.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
